File: mirea/contests/2021-09-26 MIREA Welcome Contest 2021/NoPoint C++/Задача H O(1).cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
typedef pair<double, double> pdd;
typedef pair<int, int> pii;
typedef tuple<double, double, double> td;
#define int ll


void guess(pii ans) {
  cout << "! " << ans.first << ' ' << ans.second << endl;
}

pdd ask(pii point) {
  cout << "? " << point.first << ' ' << point.second << endl;
  cout.flush();
  pdd res;
  cin >> res.first >> res.second;
  return res;
}

int32_t main() {
  const double EPS = 1e-9;
  cout << setprecision(12) << fixed;

  auto get_line = [](pdd vec, pdd from) {
    return td(vec.second, -vec.first, from.second * vec.first - from.first * vec.second);
  };
  auto get_len = [](pdd vec) {
    return sqrtl(vec.first * vec.first + vec.second * vec.second);
  };
  auto det = [](double a, double b, double c, double d) {
    return a * d - b * c;
  };
  auto intersect = [&det](td m, td n) {
    auto &[a1, b1, c1] = m;
    auto &[a2, b2, c2] = n;
    double zn = det(a1, b1, a2, b2);
    return pii(std::round(-det(c1, b1, c2, b2) / zn), std::round(-det(a1, c1, a2, c2) / zn));
  };

  // Точки one и two выбраны таким образом, чтобы на прямой,
  // построенной на этих точках в отрезке [-1e9, 1e9] были
  // только целые точки one и two
  pii one(339697208, 412598734);
  pdd vec1 = ask(one);
  if (get_len(vec1) < EPS) {
    guess(one);
    return 0;
  }
  td line1 = get_line(vec1, one);
  pii two(-579719491, 144305795);
  pdd vec2 = ask(two);
  if (get_len(vec2) < EPS) {
    guess(two);
    return 0;
  }
  td line2 = get_line(vec2, two);

  pii ans = intersect(line1, line2);
  guess(ans);
}

File: mirea/contests/2021-09-26 MIREA Welcome Contest 2021/NoPoint Python/Задача H O(1).py
from sys import stdout


def ask(point):
  print(f"? {point[0]} {point[1]}")
  stdout.flush()
  x, y = map(float, input().split())
  return x, y


def guess(point):
  print(f"! {point[0]} {point[1]}")
  stdout.flush()


def main():
  EPS = 1e-9

  def get_line(vec, point):
    return vec[1], -vec[0], point[1] * vec[0] - point[0] * vec[1]

  def get_len(vec):
    return (vec[0] * vec[0] + vec[1] * vec[1]) ** 0.5

  def det(a, b, c, d):
    return a * d - b * c

  def intersect(m, n):
    a1, b1, c1 = m
    a2, b2, c2 = n
    zn = det(a1, b1, a2, b2)
    return round(-det(c1, b1, c2, b2) / zn), round(-det(a1, c1, a2, c2) / zn)

  one = (339697208, 412598734)
  vec1 = ask(one)
  if get_len(vec1) < EPS:
    guess(one)
    return

  line1 = get_line(vec1, one)
  two = (-579719491, 144305795)
  vec2 = ask(two)
  if get_len(vec2) < EPS:
    guess(two)
    return
  line2 = get_line(vec2, two)

  ans = intersect(line1, line2)
  guess(ans)


main()

